File: app/src/main/java/com/waveneuro/ui/session/history/SessionHistoryActivity.java
package com.waveneuro.ui.session.history;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.widget.TextView;

import androidx.lifecycle.Observer;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.button.MaterialButton;
import com.waveneuro.R;
import com.waveneuro.data.model.response.session.SessionResponse;
import com.waveneuro.ui.base.BaseActivity;
import com.waveneuro.ui.session.session.SessionCommand;
import com.waveneuro.ui.session.session.SessionViewState;

import java.sql.Timestamp;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class SessionHistoryActivity extends BaseActivity {
    
    public static final String START_SESSION = "start_session";

    @Inject
    SessionHistoryViewModel sessionHistoryViewModel;

    @BindView(R.id.rvSessions)
    RecyclerView rvSessions;

    protected SessionListAdapter mAdapter;
    protected RecyclerView.LayoutManager mLayoutManager;

    int userId;

    @Inject
    SessionCommand sessionCommand;

    @OnClick(R.id.btn_start_session)
    public void onClickStartSession() {
        Intent resultIntent = new Intent();
        resultIntent.putExtra(START_SESSION, true);
        setResult(Activity.RESULT_OK, resultIntent);
        finish();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {

        activityComponent().inject(this);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_session_history);
        ButterKnife.bind(this);

        if(getIntent().hasExtra(SessionHistoryCommand.USER_ID)) {
            userId = Integer.valueOf(getIntent().getStringExtra(SessionHistoryCommand.USER_ID));
            ((TextView)findViewById(R.id.tv_name)).setText(getIntent().getStringExtra(SessionHistoryCommand.NAME));
        }

        if(getIntent().hasExtra(SessionHistoryCommand.TREATMENT_DATA_PRESENT)) {
            ((MaterialButton)findViewById(R.id.btn_start_session)).setEnabled(getIntent().getBooleanExtra(SessionHistoryCommand.TREATMENT_DATA_PRESENT, false));
        }

        setView();
        setObserver();

        sessionHistoryViewModel.getSessionHistory(userId);
    }

    private void setView() {
        mLayoutManager = new LinearLayoutManager(this);
        rvSessions.setLayoutManager(mLayoutManager);
    }

    private void setObserver() {
        this.sessionHistoryViewModel.getData().observe(this, viewStateObserver);
    }

    @OnClick(R.id.iv_back)
    public void OnClickBack() {
        onBackPressed();
    }

    Observer<SessionResponse> viewStateObserver = viewState -> {

        List<Session> sessions = new ArrayList<>();
        for (int i = 0; i < viewState.getSessions().size(); i++) {
            String pattern = "dd/MM/yyyy";
            SimpleDateFormat simpleDateFormat = new SimpleDateFormat(pattern);

            SessionResponse.Session session = viewState.getSessions().get(i);

            String dateRd;
            if (session.eegRecordedAt == null) {
                dateRd = "";
            } else {
                Timestamp timestamp = new Timestamp(Math.round(session.eegRecordedAt)*1000);
                dateRd = simpleDateFormat.format(timestamp.getTime());
            }

            String dateSd;
            if (session.finishedAt == null) {
                dateSd = "";
            } else {
                Timestamp timestamp = new Timestamp(Math.round(session.finishedAt)*1000);
                dateSd = simpleDateFormat.format(timestamp.getTime());
            }

            sessions.add(new Session(
                    session.getSonalId(),
                    dateRd,
                    dateSd,
                    session.isCompleted()));
        }
        mAdapter = new SessionListAdapter(sessions);
        rvSessions.setAdapter(mAdapter);
        mAdapter.notifyDataSetChanged();
    };

}

File: app/src/main/java/com/waveneuro/ui/session/history/SessionHistoryActivity.kt
package com.waveneuro.ui.session.history

import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.activity.viewModels
import androidx.core.view.isVisible
import androidx.lifecycle.Observer
import com.bumptech.glide.Glide
import com.waveneuro.databinding.ActivitySessionHistoryBinding
import com.waveneuro.ui.base.activity.BaseViewModelActivity
import com.waveneuro.ui.model.Session
import com.waveneuro.ui.session.history.adapter.SessionListAdapter
import com.waveneuro.ui.session.history.viewmodel.SessionHistoryViewModel
import com.waveneuro.ui.session.history.viewmodel.SessionHistoryViewModelImpl
import com.waveneuro.utils.ext.getAppComponent
import java.sql.Timestamp
import java.text.SimpleDateFormat

class SessionHistoryActivity :
    BaseViewModelActivity<ActivitySessionHistoryBinding, SessionHistoryViewModel>() {

    private lateinit var adapter: SessionListAdapter

    private var userId = 0

    override val viewModel: SessionHistoryViewModelImpl by viewModels {
        getAppComponent().sessionHistoryViewModelFactory()
    }

    override fun initBinding(): ActivitySessionHistoryBinding =
        ActivitySessionHistoryBinding.inflate(layoutInflater)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (intent.hasExtra(USER_ID)) {
            userId = intent.getIntExtra(USER_ID, 0)
        }
        if (intent.hasExtra(TREATMENT_DATA_PRESENT)) {
            binding.btnStartSession.isEnabled =
                intent.getBooleanExtra(TREATMENT_DATA_PRESENT, false)
        }

        setView()
        setObserver()

        viewModel.processEvent(SessionHistoryViewEvent.Start(userId))
    }

    private fun setView() {
        with(binding) {
            val firstName = intent.getStringExtra(FIRST_NAME) ?: ""
            val lastName = intent.getStringExtra(LAST_NAME) ?: ""
            val url = intent.getStringExtra(USER_URL) ?: ""

            tvName.text = "$firstName $lastName"
            btnStartSession.setOnClickListener {
                setResult(RESULT_OK, Intent().putExtra(START_SESSION, true))
                finish()
            }
            ivBack.setOnClickListener { onBackPressed() }

            if (url.isBlank()) {
                val initials = "${firstName[0]}${lastName[0]}".trim().uppercase()

                tvInitials.text = initials
                tvInitials.isVisible = true
            } else {
                Glide.with(this@SessionHistoryActivity)
                    .load(url)
                    .into(binding.ivProfileImage)
            }
        }
    }

    private fun setObserver() {
        viewModel.viewEffect.observe(this, Observer { viewEffect ->
            when(viewEffect) {
                is SessionHistoryViewEffect.Success -> {
                    val sessions: MutableList<Session> = mutableListOf()
                    viewEffect.sessionList.sessions.forEach { session ->
                        val pattern = "dd/MM/yyyy"
                        val simpleDateFormat = SimpleDateFormat(pattern)
                        val dateRd = session.eegRecordedAt.let {
                            val timestamp = Timestamp(Math.round(it) * 1000)
                            simpleDateFormat.format(timestamp.time)
                        } ?: ""
                        val dateSd = session.finishedAt.let {
                            val timestamp = Timestamp(Math.round(it) * 1000)
                            simpleDateFormat.format(timestamp.time)
                        } ?: ""

                        sessions.add(Session(session.sonalId, dateRd, dateSd, session.isCompleted))
                    }

                    adapter = SessionListAdapter(this, sessions)
                    binding.rvSessions.adapter = adapter
                    adapter.notifyDataSetChanged()
                }
            }
        })
    }

    companion object {
        const val START_SESSION = "start_session"

        private const val USER_ID = "user_id"
        private const val USER_URL = "user_url"
        private const val FIRST_NAME = "first_name"
        private const val LAST_NAME = "last_name"
        private const val TREATMENT_DATA_PRESENT = "treatment_data_present"

        fun newIntent(context: Context, userId: Int?, userUrl: String?,
                     firstName: String?, lastName: String?, treatmentDataPresent: Boolean) =
            Intent(context, SessionHistoryActivity::class.java).apply {
                putExtra(USER_ID, userId)
                putExtra(USER_URL, userUrl)
                putExtra(FIRST_NAME, firstName)
                putExtra(LAST_NAME, lastName)
                putExtra(TREATMENT_DATA_PRESENT, treatmentDataPresent)
            }
    }

}